client/src/util/auth.jsx

```diff
@@ -30,243 +42,228 @@ function useAuthProvider() {
-    return (
-      supabase.auth
-        .signInWithOAuth({
-          provider: name,
-          options: {
-            redirectTo: `${window.location.origin}/dashboard`,
-          },
-        })
-        .then(handleError)
-        // Because `signInWithOAuth` resolves immediately we need to add this so that
-        // it never resolves (component will display loading indicator indefinitely).
-        // Once social signin is completed the page will redirect to value of `redirectTo`.
-        .then(() => {
-          return new Promise(() => null);
-        })
-    );
+  const signin = async (email, password) => {
+    try {
+      const userCredential = await signInWithEmailAndPassword(
+        auth,
+        email,
+        password,
+      );
+
+      await handleAuth(userCredential.user);
+    } catch (error) {
+      handleError(error);
+    }
+    // return supabase.auth
+    //   .signInWithPassword({ email, password })
+    //   .then(handleError)
+    //   .then(handleAuth);
   };
 
-  const signinWithMagicLink = (email) => {
-    return supabase.auth
-      .signInWithOtp({
-        email,
-        options: {
-          emailRedirectTo: `${window.location.origin}/dashboard`,
-        },
-      })
-      .then(handleError);
+  const signinWithProvider = async (name) => {
+
+    let provider;
+
+    switch (name) {
+      case 'google':
+        provider = new GoogleAuthProvider();
+        break;
+      case 'github':
+        provider = new GithubAuthProvider();
+        break;
+      default:
+        throw new Error(`Provider ${name} is not supported`);
+    }
+
+    try {
+      await signInWithPopup(auth, provider);
+
+      window.location.href = `${window.location.origin}/dashboard`;
+    } catch (error) {
+      console.log('Error signing in with provider', error.message);
+      handleError(error);
+    }
+
+    return new Promise(() => null);
+
+    // return (
+    //   supabase.auth
+    //     .signInWithOAuth({
+    //       provider: name,
+    //       options: {
+    //         redirectTo: `${window.location.origin}/dashboard`,
+    //       },
+    //     })
+    //     .then(handleError)
+    //     // Because `signInWithOAuth` resolves immediately we need to add this so that
+    //     // it never resolves (component will display loading indicator indefinitely).
+    //     // Once social signin is completed the page will redirect to value of `redirectTo`.
+    //     .then(() => {
+    //       return new Promise(() => null);
+    //     })
+    // );
   };
 
-  const signout = () => {
-    return supabase.auth.signOut().then(handleError);
+  // const signinWithMagicLink = (email) => {
+  //   return supabase.auth
+  //     .signInWithOtp({
+  //       email,
+  //       options: {
+  //         emailRedirectTo: `${window.location.origin}/dashboard`,
+  //       },
+  //     })
+  //     .then(handleError);
+  // };
+
+  const signout = async () => {
+    signOut(auth)
```

Comment:
Should you do `return signOut(auth)` in case the signOut returns something important?

Reply:
signOut as per firebase doesnt appear to return anything: 

    import { getAuth, signOut } from "firebase/auth";
    
    const auth = getAuth();
    signOut(auth).then(() => {
      // Sign-out successful.
    }).catch((error) => {
      // An error happened.
    });

I updated the code logic to redirect user to signin page when signOut(auth) is successful

Reply:
I checked the docs, signOut returns a Promise that resolves to void. That's a bit different.

If you do `return signOut(auth)` or `await signOut(auth)`, then the caller calling this function can use `await` or Promise chaining on this function. As of right now, neither is possible because this function finishes executing and doesn't wait for the Firebase signOut code to finish

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#chaining
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await

Reply:
I recommend doing the following:
```
const signout = () => {
    return signOut(auth)
    .then(() => {
        window.location.href = `${window.location.origin}/auth/signin`
      })
      .catch((error) => {
        throw error
      });
  };
```
Or
```
const signout = async () => {
    await signOut(auth)
    window.location.href = `${window.location.origin}/auth/signin`
}
```

Reply:
updated to: 
   const signout = async () => {
      await signOut(auth)
      window.location.href = `${window.location.origin}/auth/signin`
  }